drivers/docker/reconciler_test.go

```diff
@@ -0,0 +1,150 @@
+	path := "./test-resources/docker/reconciler_containers_list.json"
+
+	f, err := os.Open(path)
+	if err != nil {
+		return
+	}
+
+	err = json.NewDecoder(f).Decode(&sampleContainerList)
+	if err != nil {
+		return
+	}
+
+	sampleNomadContainerListItem = sampleContainerList[0]
+	sampleNonNomadContainerListItem = sampleContainerList[1]
+}
+
+func Test_HasMount(t *testing.T) {
+	require.True(t, hasMount(sampleNomadContainerListItem, "/alloc"))
+	require.True(t, hasMount(sampleNomadContainerListItem, "/data"))
+	require.True(t, hasMount(sampleNomadContainerListItem, "/secrets"))
+	require.False(t, hasMount(sampleNomadContainerListItem, "/random"))
+
+	require.False(t, hasMount(sampleNonNomadContainerListItem, "/alloc"))
+	require.False(t, hasMount(sampleNonNomadContainerListItem, "/data"))
+	require.False(t, hasMount(sampleNonNomadContainerListItem, "/secrets"))
+	require.False(t, hasMount(sampleNonNomadContainerListItem, "/random"))
+}
+
+func Test_HasNomadName(t *testing.T) {
+	require.True(t, hasNomadName(sampleNomadContainerListItem))
+	require.False(t, hasNomadName(sampleNonNomadContainerListItem))
+}
+
+func TestHasEnv(t *testing.T) {
+	envvars := []string{
+		"NOMAD_ALLOC_DIR=/alloc",
+		"NOMAD_ALLOC_ID=72bfa388-024e-a903-45b8-2bc28b74ed69",
+		"NOMAD_ALLOC_INDEX=0",
+		"NOMAD_ALLOC_NAME=example.cache[0]",
+		"NOMAD_CPU_LIMIT=500",
+		"NOMAD_DC=dc1",
+		"NOMAD_GROUP_NAME=cache",
+		"NOMAD_JOB_NAME=example",
+		"NOMAD_MEMORY_LIMIT=256",
+		"NOMAD_NAMESPACE=default",
+		"NOMAD_REGION=global",
+		"NOMAD_SECRETS_DIR=/secrets",
+		"NOMAD_TASK_DIR=/local",
+		"NOMAD_TASK_NAME=redis",
+		"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
+		"GOSU_VERSION=1.10",
+		"REDIS_VERSION=3.2.12",
+		"REDIS_DOWNLOAD_URL=http://download.redis.io/releases/redis-3.2.12.tar.gz",
+		"REDIS_DOWNLOAD_SHA=98c4254ae1be4e452aa7884245471501c9aa657993e0318d88f048093e7f88fd",
+	}
+
+	require.True(t, hasEnvVar(envvars, "NOMAD_ALLOC_ID"))
+	require.True(t, hasEnvVar(envvars, "NOMAD_ALLOC_DIR"))
+	require.True(t, hasEnvVar(envvars, "GOSU_VERSION"))
+
+	require.False(t, hasEnvVar(envvars, "NOMAD_ALLOC_"))
+	require.False(t, hasEnvVar(envvars, "OTHER_VARIABLE"))
+}
+
+func TestDanglingContainerRemoval(t *testing.T) {
+	if !tu.IsCI() {
+		t.Parallel()
+	}
+	testutil.DockerCompatible(t)
+
+	task, cfg, _ := dockerTask(t)
+	require.NoError(t, task.EncodeConcreteDriverConfig(cfg))
+
+	client, d, handle, cleanup := dockerSetup(t, task)
+	defer cleanup()
+	require.NoError(t, d.WaitUntilStarted(task.ID, 5*time.Second))
+
+	c, err := client.CreateContainer(docker.CreateContainerOptions{
+		Name: "mytest-image-" + uuid.Generate(),
+		Config: &docker.Config{
+			Image: cfg.Image,
+			Cmd:   append([]string{cfg.Command}, cfg.Args...),
+		},
+	})
+	require.NoError(t, err)
+	defer client.RemoveContainer(docker.RemoveContainerOptions{
+		ID:    c.ID,
+		Force: true,
+	})
+
+	err = client.StartContainer(c.ID, nil)
+	require.NoError(t, err)
+
+	time.Sleep(1 * time.Second)
```

Comment:
Can we wait on the client instead of introducing this timer? The rest of this test avoids all timing issues which is really nice and solid.

Reply:
The sleep is mostly for introducing some delay for testing the untracked behavior rather than waiting for a condition to be met.